### PR TITLE
added button notify characteristic #566

### PR DESCRIPTION
**Pull request by kevvz**
#566 Added button and button characteristic

How to test:

Download the uf2 file for this repository and flash it onto your device. It should be in Friend/firmware/firmware_v1.0/button_build. If you can’t find it then you can generate it using the nrf SDK. I used version 2.5.0 for both
Download the updated app on your phone. Make sure to enable debug mode. Go to the debug console.
Wait until your device connects to the app (blue light.) There are two modes: press for a little bit, then press for about 2 seconds. You should get the notifications Button Pressed and Button long Press, indicating a button tap and Long press respectively.
If for some reason you can’t use the friend debugger, then you can simply open the BLE app and press the button described in 3. You should be getting the count.

Quick rundown on how it works

An interrupt service routine is attached to pin D5 according to the Xiao sense diagram. The interrupt triggers when it detects a button press, both pressed up and down. Pin D4 is configured as supply power. 
When the interrupt triggers, a function is called that checks if the button press is actually a button press (for this implementation, it rejects inputs that appear jittery, or swing in voltage too often. That way only “stable” inputs are considered pressed). Then the input is recorded. 
A work order occasionally checks the input value. Then we grab the current state of the button, whether it be idle, single pressed, double press, or in a grace period. Depending on the state and if the current GPIO pin reads 1 or 0, it will perform different actions such as notify taps, double taps, presses, etc. I may write a guide later but TLDR it is a finite state machine, that computes an action given its state and event.

Some notes on the button:

I noticed that “fat fingering” the button, which means clicking the button with your skin tends to make the button act unpredictably. The most likely reason may be the contacts on the right side of the button, in which your fingers could accidentally connect the metal and trigger the button. Try to use your fingernails or cover the button with insulation. I imagine that in future versions that you would press the button with some insulation to protect against this (much like Avi’s Friend pendant.) 
I’ve probably pressed the button about 1000 times, and the solder connecting the button is starting to break. The boards appear to be separating. 

What can you build?

Some ideas that were thrown around was attaching APIs according to button events. Also mentioned earlier was a “record” button that could simulate other necklaces on the market. Ideas for implementation are plenty!

### CODE REVIEW

**Comment by josancamon19** [comment]
Nice, what is missing for it to be stable, can you post here the uf2 file until this point?

Can we detect push down + release button? for testing a long press?

**Comment by vincentkoc** [comment]
Not sure this will work without overlay pin/device config changes in configuration. Keen to see if this works out of the box without cmake header changes

**Comment by kevvz** [comment]
I used the nrf SDK and followed the instructions on the friend firmware guide to build, and it seems to work for me, so maybe try using that.

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 1]
```diff
@@ -66,6 +70,34 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+static struct bt_uuid_128 button_uuid = BT_UUID_INIT_128(BT_UUID_128_ENCODE(0x00000003,0x0000,0x1000,0x7450,0xBE2E44B06B00));
+static struct bt_uuid_128 button_uuid_x = BT_UUID_INIT_128(BT_UUID_128_ENCODE(0x00000004,0x0000,0x1000,0x7450,0xBE2E44B06B00));
+
+static struct bt_gatt_attr button_service_attr[] = {
+    BT_GATT_PRIMARY_SERVICE(&button_uuid),
+    BT_GATT_CHARACTERISTIC(&button_uuid_x.uuid, BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ, button_data_read_characteristic, NULL, NULL),
+    BT_GATT_CCC(button_ccc_config_changed_handler, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+};
+
+static struct bt_gatt_service button_service = BT_GATT_SERVICE(button_service_attr);
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Does it have to be in a separate service? We can add this to the main service for simplicity.

**Comment by francip** [comment]
I'd suggest we add the logic for determining click | long press | double-click in the firmware, and instead of sending a counter, we send explicit events. Something like this:

```
00 - neutral
01 - button down
02 - button up
03 - click (sent after button up, if button up within specific time from button down, and no previous click within specific timeout)
04 - double click (sent after button up, if button up within specific time from button down, and there was a previous click within specific timeout)
05 - long press (send after button up, if button up later than the regular click time period from button down)
```

This way the app can base it's logic on user behavior instead, and leave the hardware events interpretation to the device. (Incidentally, this would make it easy for us to mimic button with other in future hardware revisions)

Also, while there's no second button yet, I'd suggest we make the property two bytes, and use the second byte as a button index, which would allow us to easily add more buttons down the road if necessary.

**Review by josancamon19 — CHANGES_REQUESTED** [review]
Pls check @francip recommendations.
+ convert this from draft PR to PR.

**Comment by kevvz** [comment]
> I'd suggest we add the logic for determining click | long press | double-click in the firmware, and instead of sending a counter, we send explicit events. Something like this:
> 
> ```
> 00 - neutral
> 01 - button down
> 02 - button up
> 03 - click (sent after button up, if button up within specific time from button down, and no previous click within specific timeout)
> 04 - double click (sent after button up, if button up within specific time from button down, and there was a previous click within specific timeout)
> 05 - long press (send after button up, if button up later than the regular click time period from button down)
> ```
> 
> This way the app can base it's logic on user behavior instead, and leave the hardware events interpretation to the device. (Incidentally, this would make it easy for us to mimic button with other in future hardware revisions)
> 
> Also, while there's no second button yet, I'd suggest we make the property two bytes, and use the second byte as a button index, which would allow us to easily add more buttons down the road if necessary.

Yes, that was my original intention with sending the counter to the device. I was thinking of either having the app interpret the counter as it pleased, with the exception that the hardware would send the long event even if the button was pressed down. I didn't really think anything special should happen when just pressing down or up, as the user would release it anyways expecting something to happen.  I do think it is a good idea to implement the double tap, and long press events on the hardware side, and I will get to that part later.

The second button idea is nice, but I would need to know which GPIO pins to use, and seeing as how a ton of pins are being used for the storage / speaker, I think we need to consider the state of the device after all the features we want are implemented before adding more.

**Comment by francip** [comment]
> The second button idea is nice, but I would need to know which GPIO pins to use
Nah, don't make it all the way to the actual hardware. Just make the BLE property have a second byte in its value with the index of the button (which currently will be always 0), that if we make a two-button hardware, we can change accordingly. This way you make it easy to the app to understand a protocol that supports more than one buttons, even if the current hardware iteration only has one.

> Yes, that was my original intention with sending the counter to the device.
The counter is too close to the raw HW. Depending on the user experience we want to build, the app might change its UI state based on whether the button is down, or up (hence the explicit down and up events), separately from the action that happens on click/double-click/long click. And if tomorrow we change this button to capacitive touch or a touch screen, we can change the firmware logic how the events are recognized, while still keeping the same app logic, as it doesn't care what the underlying HW implementation is.

**Comment by francip** [comment]
Basically, make it so the app developers don't have to think much about the hardware. @josancamon19 would appreciate it if you make his life easier. :-)

**Comment by kevvz** [comment]
I won't exactly make give it two button support, but in thinking about what you've suggested, I realized how unreadable my code is, so I have rewritten the code to make it much more readable in case anyone wants to add new button gestures and more buttons. I've also included the double tap feature, although you have to press it a little slow for now. If you follow the instructions above and try to double tap, single tap, long press etc then you can see it in the debug console. The hardware only sends events now too.
And button press and undress

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 2]
```diff
@@ -143,34 +143,153 @@ void button_pressed(const struct device *dev, struct gpio_callback *cb,
...
-        consecutive_hold_count = 0;
+        if (state_ == 1) {
+
+            inc_count_1++; //button is pressed
+
+            if (inc_count_1 > 50) {
+                //If button is pressed for a long time.......
+                final_button_state = 3;
+                bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));
+                printk("long tap\n");
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/src/transport.c]: nitpick - change to LOG similar to what #590 did.

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 3]
```diff
@@ -143,34 +143,153 @@ void button_pressed(const struct device *dev, struct gpio_callback *cb,
...
+
+             }
+
+
+        }
+        else if (state_ == 1 ) {
+            inc_count_1++;
+
+            if (inc_count_1 > threshold) {
+                final_button_state = 2;
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Can you change the internal state constants to macros? So instead of 2, it actually reads what the state is. Will make it easier to maintain the logic down the road.

File: app/lib/pages/home/page.dart [hunk 4]
```diff
@@ -589,6 +589,7 @@ class _HomePageWrapperState extends State<HomePageWrapper> with WidgetsBindingOb
                                         builder: (c) => ConnectedDevice(
                                               device: _device!,
                                               batteryLevel: batteryLevel,
+                                              buttonLevel: buttonLevel
```
Comment (francip) [inline app/lib/pages/home/page.dart]: Maybe buttonState?

File: app/lib/utils/ble/communication.dart [hunk 5]
```diff
@@ -259,12 +259,18 @@ Future<StreamSubscription<List<int>>?> getBleButtonLevelListener(
       onButtonLevelChange!(value[0]);
      // debugPrint('Button level at the end: ${value[0]}');
 
-      if (value[0] < 20 && value[0] > 0) {
-        debugPrint('Button pressed');
+      if (value[0] == 1) {
```
Comment (francip) [inline app/lib/utils/ble/communication.dart]: Yay! Love how simple the app logic is now. :-)

File: Friend/firmware/firmware_v1.0/src/transport.c (hunk 1 above)
Comment (kevvz) [reply Friend/firmware/firmware_v1.0/src/transport.c]: The button states don't appear to transfer when I add the characteristic to the audio service. I think space is an issue since the audio notify takes a lot of space? I think it's best if we leave this in a separate service for now.

**Review by kevvz — COMMENTED** [review]
polished

**Comment by francip** [comment]
I was looking at the firmware_1.5 and how that code deals with buttons, and I noticed it uses k_work_reschedule to deal with the delays instead of the way you are doing it (checking every 40ms). Not sure which way is better though.

**Comment by kevvz** [comment]
> I was looking at the firmware_1.5 and how that code deals with buttons, and I noticed it uses k_work_reschedule to deal with the delays instead of the way you are doing it (checking every 40ms). Not sure which way is better though.

Hm, I think our implementations are pretty similar except for the 15 ms delay instead of my 40.  As long as they work!

**Comment by josancamon19** [comment]
@francip @koconder are we ready to merge this?

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 6]
```diff
@@ -453,10 +698,54 @@ int transport_start()
...
+		LOG_INF("Error setting up D5 Pin\n");
+		return 0;
+	}
+	else {
+		LOG_INF("D5 ready\n");
+	}
+	err2 =  gpio_pin_interrupt_configure_dt(&d5_pin_input,GPIO_INT_EDGE_BOTH);
+
+	if (err2 != 0) {
+		LOG_INF("D5 unable to detect button presses\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 7]
```diff
@@ -453,10 +698,54 @@ int transport_start()
...
+	}
+
+	int err2 = gpio_pin_configure_dt(&d5_pin_input,GPIO_INPUT);
+
+	if (err2 != 0) {
+		LOG_INF("Error setting up D5 Pin\n");
+		return 0;
+	}
+	else {
+		LOG_INF("D5 ready\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Remove all the `\n` in all the logs, this is not required

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 8]
```diff
@@ -453,10 +698,54 @@ int transport_start()
...
+		LOG_INF("D5 Pin ready\n");
+	}
+	else {
+		LOG_INF("D5 Pin not ready\n");
+	}
+
+	int err2 = gpio_pin_configure_dt(&d5_pin_input,GPIO_INPUT);
+
+	if (err2 != 0) {
+		LOG_INF("Error setting up D5 Pin\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 9]
```diff
@@ -453,10 +698,54 @@ int transport_start()
...
+		LOG_INF("Error setting up D4 Pin Voltage\n");
+	}
+	else {
+		LOG_INF("D4 ready to transmit voltage\n");
+	}
+	if (gpio_is_ready_dt(&d5_pin_input)) {
+		LOG_INF("D5 Pin ready\n");
+	}
+	else {
+		LOG_INF("D5 Pin not ready\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 10]
```diff
@@ -453,10 +698,54 @@ int transport_start()
...
 
+    	if (gpio_is_ready_dt(&d4_pin)) {
+		LOG_INF("D4 Pin ready\n");
+	}
+    	else {
+		LOG_INF("Error setting up D4 Pin\n");
+	}
+
+	if (gpio_pin_configure_dt(&d4_pin, GPIO_OUTPUT_ACTIVE) < 0) {
+		LOG_INF("Error setting up D4 Pin Voltage\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 11]
```diff
@@ -203,6 +445,8 @@ static void _transport_connected(struct bt_conn *conn, uint8_t err)
         return;
     }
 
+    LOG_INF("bluetooth activated\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Type case, `LOG_INF("Bluetooth activated\n");`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 12]
```diff
@@ -89,6 +120,216 @@ struct bt_conn *current_connection = NULL;
...
+
+static void notify_long_tap() {
+    final_button_state[0] = LONG_TAP; //button press
+    LOG_INF("long tap\n");
+    bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));  
+}
+
+#define LONG_PRESS_INTERVAL 50
+#define SINGLE_PRESS_INTERVAL 2
+void check_button_level(struct k_work *work_item) {
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Move logic to a button.c and button.h too much logic here to be inside transport.c

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 13]
```diff
@@ -50,6 +75,8 @@ static struct bt_gatt_attr audio_service_attr[] = {
     BT_GATT_CHARACTERISTIC(&audio_characteristic_data_uuid.uuid, BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ, audio_data_read_characteristic, NULL, NULL),
     BT_GATT_CCC(audio_ccc_config_changed_handler, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
     BT_GATT_CHARACTERISTIC(&audio_characteristic_format_uuid.uuid, BT_GATT_CHRC_READ, BT_GATT_PERM_READ, audio_codec_read_characteristic, NULL, NULL),
+    // BT_GATT_CHARACTERISTIC(&button_uuid_x.uuid, BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ, button_data_read_characteristic, NULL, NULL),
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Remove commented out code

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 14]
```diff
@@ -35,6 +35,31 @@ static ssize_t dfu_control_point_write_handler(struct bt_conn *conn, const struc
...
+static struct bt_gatt_attr button_service_attr[] = {
+    BT_GATT_PRIMARY_SERVICE(&button_uuid),
+    BT_GATT_CHARACTERISTIC(&button_uuid_x.uuid, BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ, button_data_read_characteristic, NULL, NULL),
+    BT_GATT_CCC(button_ccc_config_changed_handler, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+};
+
+static struct bt_gatt_service button_service = BT_GATT_SERVICE(button_service_attr);
+
+static void button_ccc_config_changed_handler(const struct bt_gatt_attr *attr, uint16_t value) {
+    LOG_INF("Handler\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Descriptive log

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf [hunk 15]
```diff
@@ -147,6 +147,6 @@ CONFIG_HEAP_MEM_POOL_SIZE=4096
 CONFIG_LSM6DSL=n
 
 # Speaker
-CONFIG_I2S=y
-CONFIG_I2S_NRFX=y
-CONFIG_SPI_NRFX=y
+# CONFIG_I2S=y
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf]: revert the changes to commenting out lines

**Review by vincentkoc — CHANGES_REQUESTED** [review]
Added some comments fyi @josancamon19 @francip

**Review by kevvz — COMMENTED** [review]
refactored!

**Comment by vincentkoc** [comment]
Looks to be in good shape to test, ill have a go with the firmware and validate.

**Comment by kodjima33** [comment]
@kevvz tested but didn't work for me

Was able to cnnect and run. Installed firmware. clicked the button in tens of different ways 

Expected behavior: I see anything in debugger thatshows I clicked the button

Current behaviour: nothing changes in debugger either on short press or long press

CleanShot 2024-08-24 at 14 10 32
CleanShot 2024-08-24 at 14 10 51

**Comment by kevvz** [comment]
> @kevvz tested but didn't work for me
> 
> Was able to cnnect and run. Installed firmware. clicked the button in tens of different ways
> 
> Expected behavior: I see anything in debugger thatshows I clicked the button
> 
> Current behaviour: nothing changes in debugger either on short press or long press
> 
> CleanShot 2024-08-24 at 14 10 32 CleanShot 2024-08-24 at 14 10 51

Ah yes, I ran into the same issue when showing this to others. If you want you can use BLE scanner to see the events.

**Review by josancamon19 — APPROVED** [review]

**Review by josancamon19 — APPROVED** [review]
